### ISSUE TITLE
collect_submodules fail to find, generated spec is wrong

### ISSUE THREAD

**Issue by etern** [pyinstaller/pyinstaller#6294]
## Description of the issue
`--collect-submodules` will generate following code in the **very beginning** of spec:
```python
hiddenimports = []
hiddenimports += collect_submodules('mypackage')
```
`collect_submodules()` will always return empty list, because `sys.path` is **not yet set**, `--paths` not working here.

`sys.path` is set in `Analysis()`, which is **too late**

### Context information (for bug reports)

* Output of `pyinstaller --version`: ```4.5```
* Version of Python: 3.8
* Platform: linux
* How you installed Python: miniconda

**Comment by bwoodsend** [comment]
Hmmph, well spotted. I guess as a temporary workaround you can add to `sys.path` directly (personally I don't think we should have a `--paths` option at all) using:
```bash
PYTHONPATH='/some/extra/folder/' pyinstaller --collect-submodules=mypackage your-code.py
```
---
As for a more permanent fix, obviously the *register extra sys.path entries* bit needs to come first in the spec file. But first we'd need to decide whether or not we're happy with `pathex` being global like it is now (currently stored in `PyInstaller.conf.CONF["pathex"]`) instead of it being either per-analysis, like `pathex`'s being a parameter to the `Analysis()` class insinuates, or per spec file. If we are, then I guess the answer is to make `--paths=/some/folder` equivalent to setting `PyInstaller.conf.CONF["pathex"]` which I suppose is easy enough (although the lack of per-build vs global scoping here makes me want to puke):
```python
from PyInstaller.conf import CONF

CONF["pathex"].append("/some/folder")
```

If we **aren't** happy with its being global then we're going to have to pull up quite a lot of floorboards. We'd need some form of context object containing the additional path entries which would have to be passed around to pretty much every function throughout the spec file so whilst it's probably the most *correct* answer, I don't want to go down that path.

**Comment by etern** [comment]
Is it appropriate to pass `collect-submodules` into `Analysis()`, and collecting only after `sys.path` is set?

**Comment by rokm** [comment]
I don't think we can actually fully solve/fix this with the spec file design we currently have. There are in fact two problematic cases:
* there are packages that are located in non-standard path. These require `PYTHONPATH` being set for the unfrozen program to run (or run-time `sys.path` manipulation). Conversely, the path needs to be added to PyInstaller via `--paths` (for hooks to work; calling hookutils functions from spec file does not work because those paths are added to search path once `Analysis` is instantiated).
* there are packages that are located next to the entry-point script. When running unfrozen, these do not require any extra modifications because python implicitly searches this location. PyInstaller also dynamically adds the location of entry-point script to the search path, although that also happens only once `Analysis` is instantiated (so hooks work, but calling hookutils from spec file does not)

---

I think the fix @bwoodsend is proposing takes care only of the first case, which I think is less common (because if PYTHONPATH needs to be set for unfrozen script, it is more likely to be properly set when PyInstaller is ran). 

Solving the second one is more problematic, because the path to entry point script is passed directly to Analysis, so until that is instantiated, we don't know where to search (we *could* modify spec template to store the path to script(s) in a variable and then use that to adjust `sys.path` at the top of spec file). That's why no hookutils function (either enabled via command like switches, or used directly in the manually-written spec) works in such case. 

So the first work-around is for user to ensure that the path is in PYTHONPATH (or modify `sys.path` at the top of spec file). The second is for user to move the use of hookutils from their spec file into proper hooks for their custom packages (which, in my opinion, should be preferable anyway, as it gives you much cleaner spec).
